pgjdbc/src/main/java/com/yugabyte/ysql/LoadBalanceService.java

```diff
@@ -314,28 +395,29 @@ private static Connection getConnection(LoadBalanceProperties loadBalancePropert
   }
 
   /**
-   * @param loadBalanceProperties
+   * @param key
    * @param lb LoadBalancer instance
    * @return true if the refresh was not required or if it was successful.
    */
-  private static synchronized boolean checkAndRefresh(LoadBalanceProperties loadBalanceProperties,
+  private static synchronized String checkAndRefresh(LoadBalanceProperties.LoadBalancerKey key,
       LoadBalancer lb) {
-    if (needsRefresh(lb.getRefreshListSeconds())) {
-      Properties props = loadBalanceProperties.getOriginalProperties();
-      String url = loadBalanceProperties.getStrippedURL();
-      Properties properties = new Properties(props);
+    String uuid = null;
+    if (lb != null) {
+      uuid = lb.getUuid();
+    }
+    if (needsRefresh(lb.getRefreshListSeconds(), lb)) {
+      String url = key.getUrl();
+      Properties properties = new Properties(key.getProperties());
       properties.setProperty("socketTimeout", "15");
       HostSpec[] hspec = hostSpecs(properties);
-
-      ArrayList<String> hosts = getAllAvailableHosts(new ArrayList<>());
+      Connection controlConnection = null;
+      ArrayList<String> hosts = getAllAvailableHosts(lbKeyToUuidMap.get(key), new ArrayList<>());
       while (true) {
         boolean refreshFailed = false;
         try {
-          if (controlConnection == null || controlConnection.isClosed()) {
-            controlConnection = new PgConnection(hspec, properties, url);
-          }
+          controlConnection = new PgConnection(hspec, properties, url);
```

Comment:
I think we should avoid creating control connection for each refresh attempt. Please ensure to use existing control connection here and create a new one if not available.
Also, close the newly created control connection from within `refresh()` method outside the `while (rs.next())` loop, if `lb.getUuid() == null && uuidToClusterInfoMap.containsKey(uuid)`.